lib/cadet/updater/xml_parser.ex

```diff
@@ -0,0 +1,304 @@
+        type: ~x"./@type"o |> transform_by(&process_charlist/1),
+        max_grade: ~x"./@maxgrade"oi,
+        entity: ~x"."
+      )
+      |> Enum.map(fn param ->
+        with {:no_missing_attr?, true} <-
+               {:no_missing_attr?, not is_nil(param[:type]) and not is_nil(param[:max_grade])},
+             question when is_map(question) <- process_question_by_question_type(param),
+             question when is_map(question) <-
+               process_question_library(question, default_library),
+             question when is_map(question) <- Map.delete(question, :entity) do
+          question
+        else
+          {:no_missing_attr?, false} ->
+            Logger.error("Missing attribute(s) on PROBLEM")
+            :error
+
+          :error ->
+            :error
+        end
+      end)
+
+    if Enum.any?(questions_params, &(&1 == :error)) do
+      :error
+    else
+      {:ok, questions_params}
+    end
+  end
+
+  @spec log_error_bad_changeset(Ecto.Changeset.t(), any()) :: :ok
+  defp log_error_bad_changeset(changeset, entity) do
+    Logger.error("Invalid #{entity} changeset. Error: #{full_error_messages(changeset.errors)}")
+
+    Logger.error("Changeset: #{inspect(changeset, pretty: true)}")
+  end
+
+  @spec process_question_by_question_type(map()) :: map() | :error
+  defp process_question_by_question_type(question) do
+    entity = question[:entity]
+
+    question[:type]
+    |> case do
+      "programming" ->
+        entity
+        |> xpath(
+          ~x"."e,
+          content: ~x"./TEXT/text()" |> transform_by(&process_charlist/1),
+          solution_template: ~x"./SNIPPET/TEMPLATE/text()" |> transform_by(&process_charlist/1),
+          solution: ~x"./SNIPPET/SOLUTION/text()" |> transform_by(&process_charlist/1),
+          autograder:
+            ~x"./SNIPPET/GRADER/text()"l
+            |> transform_by(&Enum.map(&1, fn charlist -> process_charlist(charlist) end))
+        )
+
+      "mcq" ->
+        choices =
+          entity
+          |> xpath(
+            ~x"./CHOICE"el,
+            content: ~x"./text()" |> transform_by(&process_charlist/1),
+            is_correct: ~x"./@correct"s |> transform_by(&String.to_atom/1)
+          )
+          |> Enum.with_index()
+          |> Enum.map(fn {choice, id} -> Map.put(choice, :choice_id, id) end)
+
+        entity
+        |> xpath(~x"."e, content: ~x"./TEXT/text()" |> transform_by(&process_charlist/1))
+        |> Map.put(:choices, choices)
+
+      _ ->
+        Logger.error("Invalid question type.")
+        :error
+    end
+    |> case do
+      :error -> :error
+      question_map -> Map.put(question, :question, question_map)
+    end
+  end
+
+  @spec process_question_library(map(), any()) :: map() | :error
+  defp process_question_library(question, default_library) do
+    library = xpath(question[:entity], ~x"./DEPLOYMENT"o) || default_library
+    grading_library = xpath(question[:entity], ~x"./GRADERDEPLOYMENT"o) || library
+
+    if library do
```

Comment:
If I'm reading this right, when `grading_library` is not specified, it will be a duplicate of `library` so I can just send `grading_library` to autograder?